packages/modular-scripts/src/__tests__/test.test.ts

```diff
@@ -28,45 +28,191 @@ function clearTests() {
+        try {
+          await execa(
+            'yarnpkg',
+            ['modular', 'test', 'test/ValidTest.test.ts'],
+            {
+              all: true,
+              cleanup: true,
+            },
+          );
+        } catch (error) {
+          errorNumber = (error as ExecaError).exitCode;
+        }
+        expect(errorNumber).toEqual(0);
+      });
     });
   });
 
-  describe('when the tests pass', () => {
-    it('should exit with no error', async () => {
-      let errorNumber = 0;
-      try {
-        await execa('yarnpkg', ['modular', 'test', 'test/ValidTest.test.ts'], {
-          all: true,
-          cleanup: true,
-        });
-      } catch (error) {
-        errorNumber = (error as ExecaError).exitCode;
+  describe('test command can successfully do selective tests based on the state of the repository', () => {
+    const fixturesFolder = path.join(
+      __dirname,
+      Array.from({ length: 4 }).reduce<string>(
+        (acc) => `${acc}..${path.sep}`,
+        '',
+      ),
+      '__fixtures__',
+      'ghost-testing',
+    );
+
+    const currentModularFolder = getModularRoot();
+    let randomOutputFolder: string;
+
+    beforeEach(() => {
+      // Create random dir
+      randomOutputFolder = tmp.dirSync({ unsafeCleanup: true }).name;
+      fs.copySync(fixturesFolder, randomOutputFolder);
+
+      // Create git repo & commit
+      if (process.env.GIT_AUTHOR_NAME && process.env.GIT_AUTHOR_EMAIL) {
+        execa.sync('git', [
+          'config',
+          '--global',
+          'user.email',
+          `"${process.env.GIT_AUTHOR_EMAIL}"`,
+        ]);
+        execa.sync('git', [
+          'config',
+          '--global',
+          'user.name',
+          `"${process.env.GIT_AUTHOR_NAME}"`,
+        ]);
       }
-      expect(errorNumber).toEqual(0);
+      execa.sync('git', ['init'], {
+        cwd: randomOutputFolder,
+      });
+      execa.sync('yarn', {
+        cwd: randomOutputFolder,
+      });
+      execa.sync('git', ['add', '.'], {
+        cwd: randomOutputFolder,
+      });
+      execa.sync('git', ['commit', '-am', '"First commit"'], {
+        cwd: randomOutputFolder,
+      });
+    });
+
+    // These expects run in a single test, serially for performance reasons (the setup time is quite long)
+    it('finds no unchanged using --changed / finds changed after modifying some workspaces / finds ancestors using --ancestors', () => {
+      const resultUnchanged = runRemoteModularTest(
+        currentModularFolder,
+        randomOutputFolder,
+        ['test', '--changed'],
+      );
+      expect(resultUnchanged.stdout).toContain('No changed workspaces found');
+
+      fs.appendFileSync(
+        path.join(randomOutputFolder, '/packages/b/src/index.ts'),
+        "\n// Comment to package b's source",
+      );
+      fs.appendFileSync(
+        path.join(randomOutputFolder, '/packages/c/src/index.ts'),
+        "\n// Comment to package c's source",
+      );
+
```

Comment:
Nice, funky tests :) Just a sense check: this will also work if the changes have already been committed in git, right?

Reply:
It depends; this will just compare your current branch with a specified branch; so yes, if you have committed in `your-branch` (but haven't pushed your commit) and you use `origin/your-branch` as comparison branch, the tests will run.

To frame this a bit better: by default, `--changed` compares with the default branch, so the most common case is that you're in a feature branch, you branched from the repo's default branch and you want to know what tests to execute in CI. In this case, tests will run for everything that's changed from the default branch, even if the PR branch has been committed and pushed. I didn't want this test to branch and commit for simplicity, so I'm just modifying files on the fly on the main branch and leave them uncommitted.